/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
syntax = "proto3";

package profiler.proto;
option java_package = "com.android.tools.profiler.proto";
option java_outer_classname = "Agent";

service AgentService {
  // A simple ping mechanism to notify perfd of the agent's existence, vice versa.
  rpc HeartBeat(HeartBeatRequest) returns (HeartBeatResponse) {}
}

message HeartBeatRequest {
  // The process' id. Note that the agent's pid is the same as the pid we use for
  // the heartbeat because the agent resides within the process being profiled.
  int32 pid = 1;
}

message HeartBeatResponse {
}

enum SocketType {
  UNSPECIFIED_SOCKET = 0;
  ABSTRACT_SOCKET = 1;
}

// Proto used for configuring perfd + perfa.
// The message itself is created and pushed to device from Studio
// and can be access via profiler::Config::Instance().
message AgentConfig {
  message MemoryConfig {
    // equivalent to profiler.livealloc flag in Studio.
    bool use_live_alloc = 1;
    // the maximum depth of allocation stack traces to record.
    int32 max_stack_depth = 2;
    // Enable tracking of creation and deletion of globla JNI references.
    bool track_global_jni_refs = 3;
    // Path to the app dir: usually /data/app/
    string app_dir = 4;
  }

  // equivalent to profiler.jvmti flag in Studio
  bool use_jvmti = 1;
  MemoryConfig mem_config = 2;
  // Which socket type we are using when setting up our service
  // if the device is O+ we use service_socket_name and
  // ABSTRACT_SOCKET. If the device is pre O we use
  // UNSPECIFIED_SOCKET and use the service_address.
  SocketType socket_type = 3;
  // address used for legacy devices (Nougat or older).
  string service_address = 4;
  // address used for jvmti devices that use unix sockets.
  string service_socket_name = 5;
  // equivalent to StudioFlags.PROFILER_NETWORK_REQUEST_PAYLOAD.
  bool profiler_network_request_payload = 6;
  // equivalent to StudioFlags.PROFILER_ENERGY_PROFILER_ENABLED.
  bool energy_profiler_enabled = 8;
  // equivalent to StudioFlags.PROFILER_CPU_API_TRACING.
  bool cpu_api_tracing_enabled = 9;
  // Android API level as defined in com.android.sdklib.AndroidVersion.
  int32 android_feature_level = 10;
}
